File: Visual Studio Projects/OverprintSimulation/CustomSpotMergingCS/CustomSpotMerge.cs
// -----------------------------------------------------------------------
// <copyright file="CustomSpotMerge.cs" company="Hybrid Software">
//   Copyright (c) 2025 Hybrid Software.
// </copyright>
// <summary>
//   Example provided "as is", without warranty of any kind.
// </summary>
// -----------------------------------------------------------------------

using System;
using System.Collections.Generic;
using System.IO;
using System.Runtime.InteropServices;
using JawsMako;

class CustomSpotMerge
{
    static void Main(string[] args)
    {
        // Adjust to your test files folder
        string testFilePath = @"..\..\..\..\TestFiles\";

        try
        {
            var mako = IJawsMako.create();
            IJawsMako.enableAllFeatures(mako);

            // Load the document
            var docAsm = IPDFInput.create(mako).open(Path.Combine(testFilePath, "Robots Plus Process Colors.pdf"));
            var doc = docAsm.getDocument();

            for (uint pageIndex = 0; pageIndex < doc.getNumPages(); pageIndex++)
            {
                var fixedPage = doc.getPage(pageIndex).getContent();

                // Page bounds & raster geometry
                var bounds = new FRect(0, 0, fixedPage.getWidth(), fixedPage.getHeight());
                double resolution = 150.0;
                uint pixelWidth = (uint)Math.Round(bounds.dX / 96.0 * resolution);
                uint pixelHeight = (uint)Math.Round(bounds.dY / 96.0 * resolution);

                // Colorspace must be CMYK for spot merging (we can convert to RGB later)
                var cmyk = IDOMColorSpaceDeviceCMYK.create(mako);

                // ----- Find inks and build spot list (names + CMYK components) -----
                var spots = IRendererTransform.inkInfoToColorantInfo(mako, IRendererTransform.findInks(mako, fixedPage), cmyk);
                var spotNames = new CEDLVectWString();

                foreach (var spot in spots.toVector())
                    spotNames.append(spot.name);

                int numProcess = cmyk.getNumComponents(); // 4 (CMYK)
                int numSpots = (int)spots.size();
                int numBuffers = numProcess + numSpots;

                // ----- Prepare frame buffers for renderSeparationsToFrameBuffers() -----
                var buffers = new byte[numBuffers][];
                var fb = new CEDLVectCFrameBufferInfo();

                for (int i = 0; i < numBuffers; i++)
                {
                    buffers[i] = new byte[pixelHeight *  pixelWidth];

                    var info = new IJawsRenderer.CFrameBufferInfo
                    {
                        bufferOfs = 0,
                        rowStride = (int)pixelWidth,   // pixels per row (1 byte per pixel)
                        pixelStride = 0                // tightly packed
                    };
                    fb.append(info);
                }

                // ----- Render true separations (process first, then spots) -----
                var renderer = IJawsRenderer.create(mako);
                renderer.renderSeparationsToFrameBuffers(
                    fixedPage,
                    8,                    // bits per component per plate
                    true,                 // antialiased
                    pixelWidth,
                    pixelHeight,
                    cmyk,                 // process space -> first 4 plates (C,M,Y,K)
                    buffers,
                    fb,
                    0,                    // bandMemorySize (0 = no banding)
                    bounds,
                    spotNames            // which spots to keep (all found)
                );

                // ----- Create writer and image -----
                var pair = IDOMRawImage.createWriterAndImage(
                    mako,
                    cmyk,
                    pixelWidth,
                    pixelHeight,
                    8,
                    resolution, resolution
                );

                IImageFrameWriter frameWriter = pair.frameWriter;

                // ----- Get spot components of spots to merge -----
                var components = new CEDLVectVectFloat();
                for (uint i = 0; i < numSpots; i++)
                    components.append(new CEDLVectFloat(new StdVectFloat()
                    {
                        spots[i].components[0],
                        spots[i].components[1],
                        spots[i].components[2],
                        spots[i].components[3]
                    }));

                // ----- Merge each spot component with the process buffer values and write the result to the scanline -----
                const float inv255 = 1.0f / 255.0f;
                for (uint y = 0; y < pixelHeight; y++)
                {
                    var rowStart = (int)(y * pixelWidth);
                    var scanline = new byte[pixelWidth * numProcess];

                    for (uint x = 0; x < pixelWidth; ++x)
                    {
                        for (uint c = 0; c < numProcess; ++c)
                        {
                            scanline[x * numProcess + c] = buffers[c][rowStart + x];
                            for (uint i = 0; i < numSpots; ++i)
                            {
                                float spotVal = buffers[numProcess + i][rowStart + x] * inv255;
                                float currentVal = scanline[x * numProcess + c] * inv255;
                                float newVal = 1.0f - (1.0f - components[i][c] * spotVal) * (1.0f - currentVal);
                                scanline[x * numProcess + c] = (byte)(newVal * 255.0f + 0.5f);
                            }
                        }
                    }
                    frameWriter.writeScanLine(scanline);
                }

                frameWriter.flushData();

                // ----- Convert to RGB (optional) and save as a JPEG (or other image type) -----
                var rgb = IDOMColorSpaceDeviceRGB.create(mako);
                var cc = IDOMImageColorConverterFilter.create(mako, rgb, eRenderingIntent.eRelativeColorimetric, eBlackPointCompensation.eBPCDefault);
                var filtered = IDOMFilteredImage.create(mako, pair.domImage, cc);

                string outJpeg = $"output_{pageIndex}.jpg";
                IDOMJPEGImage.encode(mako, filtered, IOutputStream.createToFile(mako, outJpeg));
                Console.WriteLine($"Wrote: {outJpeg}");
            }
        }
        catch (MakoException me)
        {
            Console.Error.WriteLine($"Mako error {me.m_errorCode}: {me.m_msg}");
            Environment.Exit((int)me.m_errorCode);
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine(ex);
            Environment.Exit(1);
        }
    }
}

File: Python Projects/OverprintSimulation/custom_spot_merge.py
#
# -----------------------------------------------------------------------
# <copyright file="custom_spot_merge.py" company="Hybrid Software Helix Ltd">
#  Copyright (C) 2025 Hybrid Software Helix Ltd. All rights reserved.
# </copyright>
# <summary>
#  This example is provided on an "as is" basis and without warranty of any kind.
#  Hybrid Software Helix Ltd. does not warrant or make any representations
#  regarding the use or results of use of this example.
# </summary>
# -----------------------------------------------------------------------
#

import math
import sys
import os
from pathlib import Path

import jawsmakoIF_python
from jawsmakoIF_python import *

TEST_FILES_PATH = str(Path(__file__).resolve().parents[2] / "TestFiles") + os.sep

def custom_spot_merge():
    # Adjust to your test files folder

    try:
        # Instantiate Mako
        mako = IJawsMako.create("")
        IJawsMako.enableAllFeatures(mako)
        factory = mako.getFactory()

        # Load the document
        pdf_input = IPDFInput.create(mako)
        doc_asm = pdf_input.open(TEST_FILES_PATH + "Robots Plus Process Colors.pdf")
        document = doc_asm.getDocument()

        for page_index in range(document.getNumPages()):
            fixed_page = document.getPage(page_index).getContent()

            # Page bounds & raster geometry
            bounds = FRect(0, 0, fixed_page.getWidth(), fixed_page.getHeight())
            resolution = 150.0
            pixel_width = int(round(bounds.dX / 96.0 * resolution))
            pixel_height = int(round(bounds.dY / 96.0 * resolution))

            # Colorspace must be CMYK for spot merging
            cmyk = IDOMColorSpaceDeviceCMYK.create(factory)

            # Find inks and build spot list
            inks = IRendererTransform.findInks(mako, fixed_page)
            spots = IRendererTransform.inkInfoToColorantInfo(mako, inks, cmyk)

            spot_names = CEDLVectWString()
            for i in range(spots.size()):
                spot_names.append(spots.getitem(i).name)

            num_process = cmyk.getNumComponents()  # 4
            num_spots = int(spots.size())
            num_buffers = num_process + num_spots

            # Prepare frame buffers
            buffers = []
            fb = CEDLVectCFrameBufferInfo()

            for i in range(num_buffers):
                # allocate pixel buffer for this plane
                buf = bytearray(pixel_width * pixel_height)
                buffers.append(buf)

                info = IJawsRenderer.CFrameBufferInfo()
                info.bufferOfs = 0
                info.rowStride = pixel_width
                info.pixelStride = 0
                fb.append(info)

            # Render true separations
            renderer = IJawsRenderer.create(mako)
            renderer.renderSeparationsToFrameBuffers(
                fixed_page,
                8,
                True,
                pixel_width,
                pixel_height,
                cmyk,
                buffers,
                fb,
                0,
                bounds,
                spot_names
            )

            # Create writer and image
            pair = IDOMRawImage.createWriterAndImage(
                mako,
                cmyk,
                pixel_width,
                pixel_height,
                8,
                resolution,
                resolution
            )
            frame_writer = pair.frameWriter

            # Spot components to merge
            components = []
            for i in range(num_spots):
                spot = spots.getitem(i)
                vals = spot.components
                components.append([vals.getitem(c) for c in range(num_process)])

            # Merge each spot buffer with process values
            inv255 = 1.0 / 255.0
            scanline = bytearray(pixel_width * num_process)

            for y in range(pixel_height):
                row_start = y * pixel_width
                for x in range(pixel_width):
                    for c in range(num_process):
                        idx = x * num_process + c
                        scanline[idx] = buffers[c][row_start + x]
                        for i in range(num_spots):
                            spot_val = buffers[num_process + i][row_start + x] * inv255
                            current_val = scanline[idx] * inv255
                            new_val = 1.0 - (1.0 - components[i][c] * spot_val) * (1.0 - current_val)
                            scanline[idx] = int(new_val * 255.0 + 0.5)
                frame_writer.writeScanLine(scanline)

            frame_writer.flushData()

            # Convert to RGB and save as JPEG
            rgb = IDOMColorSpaceDeviceRGB.create(factory)
            cc = IDOMImageColorConverterFilter.create(
                factory,
                rgb,
                eRelativeColorimetric,
                eBPCDefault
            )
            filtered = IDOMFilteredImage.create(factory, pair.domImage, cc)

            out_jpeg = f"output_{page_index}.jpg"
            IDOMJPEGImage.encode(mako, filtered, IOutputStream.createToFile(factory, out_jpeg))

            print(f"Wrote: {out_jpeg}")

    except MakoException as ex:
        print(f"MakoException {ex.m_errorCode}: {ex.m_msg}")
        return 1
    except Exception as ex:
        print(f"Error: {ex}")
        return 1

    return 0


if __name__ == "__main__":
    sys.exit(custom_spot_merge())
